haystack/components/preprocessors/markdown_header_splitter.py

```diff
@@ -0,0 +1,355 @@
+# SPDX-FileCopyrightText: 2022-present deepset GmbH <info@deepset.ai>
+#
+# SPDX-License-Identifier: Apache-2.0
+
+import re
+from typing import Literal, Optional
+
+from haystack import Document, component, logging
+from haystack.components.preprocessors import DocumentSplitter
+
+logger = logging.getLogger(__name__)
+
+
+@component
+class MarkdownHeaderSplitter:
+    """
+    Split documents at Markdown headers, with optional secondary splitting and header level inference.
+
+    This component processes text documents by:
+    - Splitting them into chunks at Markdown headers (e.g., '#', '##', etc.), preserving header hierarchy as metadata.
+    - Optionally inferring and rewriting header levels for documents where header structure is ambiguous.
+    - Optionally applying a secondary split (by word, passage, period, or line) to each chunk.
+      This is done in haystack's DocumentSplitter.
+    - Preserving and propagating metadata such as parent headers, page numbers, and split IDs.
+    """
+
+    def __init__(
+        self,
+        infer_header_levels: bool = False,
```

Comment:
Let's make these keyword only
```suggestion
        self,
        *,
        infer_header_levels: bool = False,
```